soundcard/utils.py

```diff
@@ -0,0 +1,21 @@
+def match_device(id, devices):
+    """Find id in a list of devices.
+
+    id can be a platfom-specific id, a substring of the device name, or a
+    fuzzy-matched pattern for the microphone name.
+
+    """
+    devices_by_id = {device.id: device for device in devices}
+    devices_by_name = {device.name: device for device in devices}
+    if id in devices_by_id:
+        return devices_by_id[id]
+    # try substring match:
+    for name, device in devices_by_name.items():
+        if id in name:
```

Comment:
On macOS, device IDs are integers (IIRC). However, `match_device` will break here if an integer is given. I think we should add a line such as `if not isinstance(id, str): raise IndexError('no device with id {}'.format(id))`.

To be clear, this is not a bug you introduced, but one preexisting in the code.

Reply:
Ouch, you are right... Integer check added, including a test.